dedalus/core/basis.py

```diff
@@ -4969,4 +5037,114 @@ def operate(self, out):
+            spacing.append(axis_spacing)
+        return spacing
+
+    def compute_cfl_frequency(self, velocity, out):
+        u_mag = np.abs(velocity.data)
+        out.data[:] = 0
+        for i, dx in enumerate(self.cfl_spacing(velocity)):
+            out.data += u_mag[i] / dx
+
+
+class D2AdvectiveCFL(operators.AdvectiveCFL):
+
+    input_coord_type = PolarCoordinates
+    input_basis_type = DiskBasis
+
+    @CachedMethod
+    def cfl_spacing(self):
+        #Assumes velocity is a 2-length vector over polar coordinates
+        basis   = self.input_basis
+        dealias = basis.dealias
+        azimuth_spacing = basis.local_grid_spacing(0, scales=dealias)
+        if basis.mmax == 0:
+            azimuth_spacing[:] = np.inf
+        else:
+            azimuth_spacing[:] = basis.radius / basis.mmax
+        radial_spacing = dealias[1] * basis.local_grid_spacing(1, scales=dealias)
+        return [azimuth_spacing, radial_spacing]
+
+    def compute_cfl_frequency(self, velocity, out):
+        #Assumes velocity is a 2-length vector over polar coordinates
+        u_theta, u_r = np.abs(velocity.data)
+        out.data[:] = u_theta / self.cfl_spacing()[0]
+        out.data += u_r / self.cfl_spacing()[1]
+
+
+class S2AdvectiveCFL(operators.AdvectiveCFL):
+
+    input_coord_type = S2Coordinates
+    input_basis_type = SWSH
+
+    @CachedMethod
+    def cfl_spacing(self, r=None):
+        #Assumes velocity is a 2-length vector over spherical coordinates
+        basis   = self.input_basis
+        dealias = basis.dealias
+        if r is None: r = basis.radius
+        s2_spacing = basis.local_grid_spacing(0, scales=dealias)
+        if basis.Lmax == 0:
+            s2_spacing[:] = np.inf
+        else:
+            s2_spacing[:] = r / np.sqrt(basis.Lmax*(basis.Lmax + 1))
+        return [s2_spacing,]
+
+    def compute_cfl_frequency(self, velocity, out):
+        #compute u_mag * sqrt(ell*(ell+1)) / r
+        #Assumes velocity is a 2-length vector over spherical coordinates
+        u_phi, u_theta = velocity.data[0], velocity.data[1]
+        u_mag = np.sqrt(u_phi**2 + u_theta**2)
+        #Again assumes that this field only has an S2 basis
+        out.data[:] = u_mag / self.cfl_spacing()[0]
+
+
+class Spherical3DAdvectiveCFL(operators.AdvectiveCFL):
+
+    input_coord_type = SphericalCoordinates
+    input_basis_type = (BallBasis, ShellBasis)
+
+    @CachedMethod
+    def cfl_spacing(self):
+        #Assumes velocity is a 3-length vector over spherical coordinates.
+        basis   = self.input_basis
+        dealias = basis.dealias
+        if isinstance(basis, BallBasis):
+            spacings = S2AdvectiveCFL.cfl_spacing(self, r=basis.radial_basis.radius)
+        elif isinstance(basis, ShellBasis):
+            spacings = S2AdvectiveCFL.cfl_spacing(self, r=1)
+            spacings[0] = spacings[0] * basis.local_grid_radius(dealias[2]) #get proper radial scaling for shell
```

Comment:
Why multiply separately instead of passing `local_grid_radiu` to `S2AdvectiveCFL.cfl_spacing`?

Reply:
cfl_spacing is a CachedMethod, and I find that passing in a numpy ndarray as a kwarg throws this error:

```
    def __call__(self, *args, **kw):
        # First try direct call signature
        direct_call = (args, tuple(kw.items()))
        try:
>           return self.cache[direct_call]
E           TypeError: unhashable type: 'numpy.ndarray'
```

I want to make sure that cfl spacings are cached to avoid unnecessary computations. Thoughts on how to handle this?

Reply:
Ah ok, yeah looks good then!